hera_mc/cm_hookup.py

```diff
@@ -173,46 +64,128 @@ def get_hookup(self, hpn, rev, port, at_date, state_args, exact_match=False):
+        """
+        conn = self.__get_next_connection(direction, part, rev, pol)
+        if conn is not None:
+            if direction == 'up':
+                self.upstream.append(conn)
+                part = conn.upstream_part
+                rev = conn.up_part_rev
+            else:
+                self.downstream.append(conn)
+                part = conn.downstream_part
+                rev = conn.down_part_rev
+            self.__recursive_go(direction, part, rev, pol)
 
+    def __get_next_connection(self, direction, part, rev, pol):
+        """
+        Get next connected part going the given direction.
+        """
+        options = []
+        if direction.lower() == 'up':      # Going upstream
+            for conn in self.session.query(PC.Connections).filter(
+                    (PC.Connections.downstream_part == part) &
+                    (PC.Connections.down_part_rev == rev)):
+                conn.gps2Time()
+                if cm_utils._is_active(self.at_date, conn.start_date, conn.stop_date):
+                    nppart = conn.upstream_part
+                    npport = conn.upstream_output_port
+                    options.append(copy.copy(conn))
+        elif direction.lower() == 'down':  # Going downstream
+            for conn in self.session.query(PC.Connections).filter(
+                    (PC.Connections.upstream_part == part) &
+                    (PC.Connections.up_part_rev == rev)):
+                conn.gps2Time()
+                if cm_utils._is_active(self.at_date, conn.start_date, conn.stop_date):
+                    nppart = conn.downstream_part
+                    npport = conn.downstream_input_port
+                    options.append(copy.copy(conn))
+        next_one = None
+        if len(options) == 0:
+            next_one = None
+        elif len(options) == 1:
+            next_one = options[0]
+        else:
+            for opc in options:
+                if direction.lower() == 'up':
+                    if self.__get_pols_to_do(opc.upstream_part, opc.upstream_output_port, check_pol=pol):
+                        next_one = opc
+                        break
+                else:
+                    if self.__get_pols_to_do(opc.downstream_part, opc.downstream_input_port, check_pol=pol):
+                        next_one = opc
+                        break
+        return next_one
+
+    def __add_hookup_timing(self, hookup_dict):
+        really_late = 9999999999
+        hookup_dict['timing'] = {}
+        for akey, hk in hookup_dict['hookup'].iteritems():
+            hookup_dict['timing'][akey] = {}
+            for pkey, pol in hk.iteritems():
+                latest_start = 0
+                earliest_stop = really_late
+                for c in pol:
+                    if c.start_gpstime > latest_start:
+                        latest_start = c.start_gpstime
+                    if c.stop_gpstime is not None and c.stop_gpstime < earliest_stop:
```

Comment:
Ok, I think I see -- the really_late value is just a starting point and it can go down from there. I think it works ok, it's just a little more confusing to read than if earliest_stop starts out as None and only gets an assignment if there is a stoptime.